From 935d732fdb1935cf0f150deaf97e735f701a948f Mon Sep 17 00:00:00 2001
From: Zheng Yejian <zhengyejian1@huawei.com>
Date: Fri, 14 Oct 2022 17:10:53 +0000
Subject: [PATCH 002/595] [Huawei] livepatch/core: Fix livepatch/state leak on
 error path

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: NA

--------------------------------

File '/proc/livepatch/state' should be removed before removing
'/proc/livepatch', otherwise following log will appear:
  remove_proc_entry: removing non-empty directory '/proc/livepatch',
  leaking at least 'state'

Fixes: c33e42836a74 ("livepatch/core: Allow implementation without ftrace")
Signed-off-by: Zheng Yejian <zhengyejian1@huawei.com>
---
 kernel/livepatch/core.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/kernel/livepatch/core.c b/kernel/livepatch/core.c
index 270da0e311ff..511f5a71e994 100644
--- a/kernel/livepatch/core.c
+++ b/kernel/livepatch/core.c
@@ -2758,13 +2758,15 @@ static int __init klp_init(void)
 
 	klp_root_kobj = kobject_create_and_add("livepatch", kernel_kobj);
 	if (!klp_root_kobj)
-		goto error_remove;
+		goto error_remove_state;
 
 #ifdef CONFIG_LIVEPATCH_STOP_MACHINE_CONSISTENCY
 	arch_klp_init();
 #endif
 	return 0;
 
+error_remove_state:
+	remove_proc_entry("livepatch/state", NULL);
 error_remove:
 	remove_proc_entry("livepatch", NULL);
 error_out:
-- 
2.17.1

